### ISSUE TITLE
Segmentation lines stuck together

### ISSUE THREAD

**Issue by johnlockejrr** [qurator-spk/eynollah#163]
Any idea what can I do so my segmented lines won't be stuck together? Any segmentation option? Or needs a new model altogether?

`eynollah layout -i BL_Add._27296/add_ms_27296_f004v.jpg -o outputs -m ./models_eynollah -slro -ncl 1 -ncu 1 -light -tll`

![Image](https://github.com/user-attachments/assets/ac2d6f7d-3f49-4470-971b-2b1b2d1d2dd4)

The full segmentation performes better but still there are lines stuck together.

`eynollah layout -i BL_Add._27296/add_ms_27296_f004v.jpg -o outputs -m ./models_eynollah -slro -ncl 1 -ncu 1`

![Image](https://github.com/user-attachments/assets/0cc12710-71bc-481b-b6fa-d005ba063af7)

**Comment by vahidrezanezhad** [comment]
@johnlockejrr  Could you please provide the input images?

**Comment by johnlockejrr** [comment]
Sure!

[BL_Add._27296_images.zip](https://github.com/user-attachments/files/20140513/BL_Add._27296_images.zip)

**Comment by vahidrezanezhad** [comment]
@johnlockejrr This issue has already been resolved and will be merged soon.

**Comment by johnlockejrr** [comment]
What can I do in this case? I added the artificial class but I'm not sure is enough for the line not to stick together.
Dataset json config:
```
{
    "use_case": "textline",
    "artificial_class_label": 2
}
```

Original image:

![Image](https://github.com/user-attachments/assets/c1bae467-7d75-4e28-bb6d-33fdcf9b2c53)

3d visualization for the label:

![Image](https://github.com/user-attachments/assets/556c2006-0a87-4bb3-9276-1c19402b8a32)

**Comment by johnlockejrr** [comment]
By the way, I trained another model (on another dataset) and I'm very satisfied so far (some lines are stuck but very few).

![Image](https://github.com/user-attachments/assets/dac7208e-019f-46cb-ae20-9489c8bcc838)

Could I improve further?

Dataset config:
```
{
    "use_case": "textline",
    "artificial_class_label": 2
}
```

Model config:

```
{
  "n_classes": 3,
  "n_epochs": 50,
  "input_height": 672,
  "input_width": 448,
  "weight_decay": 1e-06,
  "n_batch": 4,
  "learning_rate": 0.0001,
  "early_stopping_patience": 15,
  "early_stopping_monitor": "val_loss",
  "reduce_lr_patience": 7,
  "reduce_lr_factor": 0.1,
  "reduce_lr_min": 1e-07,
  "reduce_lr_monitor": "val_loss",
  "patches": false,
  "augmentation": true,
  "flip_aug": false,
  "blur_aug": false,
  "padding_white": false,
  "padding_black": false,
  "scaling": true,
  "shifting": false,
  "degrading": false,
  "brightening": false,
  "binarization": false,
  "adding_rgb_background": false,
  "adding_rgb_foreground": false,
  "add_red_textlines": false,
  "channels_shuffling": false,
  "dir_train": "/home/incognito/sbb_new/dataset/sam_41_mss/dir_train/train",
  "dir_eval": "/home/incognito/sbb_new/dataset/sam_41_mss/dir_train/eval",
  "dir_output": "runs/sam_41_mss",
  "pretraining": true,
  "scaling_bluring": false,
  "scaling_binarization": false,
  "rotation": false,
  "rotation_not_90": false,
  "scaling_brightness": false,
  "scaling_flip": false,
  "thetha": [
    10,
    -10
  ],
  "shuffle_indexes": null,
  "blur_k": [
    "blur",
    "guass",
    "median"
  ],
  "scales": [
    0.6,
    0.7,
    0.8,
    0.9
  ],
  "degrade_scales": [
    0.2,
    0.4
  ],
  "brightness": [
    1.3,
    1.5,
    1.7,
    2
  ],
  "flip_index": [
    0,
    1,
    -1
  ],
  "continue_training": false,
  "transformer_patchsize_x": null,
  "transformer_patchsize_y": null,
  "transformer_num_patches_xy": null,
  "transformer_projection_dim": 64,
  "transformer_mlp_head_units": [
    128,
    64
  ],
  "transformer_layers": 8,
  "transformer_num_heads": 4,
  "transformer_cnn_first": true,
  "index_start": 0,
  "dir_of_start_model": " ",
  "is_loss_soft_dice": true,
  "weighted_loss": false,
  "data_is_provided": false,
  "task": "segmentation",
  "f1_threshold_classification": null,
  "classification_classes_name": null,
  "backbone_type": "nontransformer",
  "save_interval": null,
  "dir_img_bin": null,
  "number_of_backgrounds_per_image": 1,
  "dir_rgb_backgrounds": null,
  "dir_rgb_foregrounds": null,
  "seed": 691512072
}
```

**Comment by vahidrezanezhad** [comment]
> [@johnlockejrr](https://github.com/johnlockejrr) The artificial class can be very helpful based on my experience working with it. However, it is most effective when the ground truth (GT) textlines are not overlapping. This is because the artificial class is generated by dilating the textline contours, and the result is then written onto the label.
> 
> In your case, I notice that the contours overlap significantly. This makes using the artificial class much more complex, as the artificial class for one contour overlaps with its neighboring textline, and vice versa. A better approach would be for the artificial class to be positioned exactly between two textlines, right at their border.
> 
> Given this, I recommend applying [erosion_early](https://github.com/qurator-spk/sbb_pixelwise_segmentation/blob/3de24243fc061d6ef6810515e82cccc4492d7646/gt_gen_utils.py#L688). Unfortunately, this parameter is currently static in the script, but I believe it should be passed as an argument for greater flexibility. You can play with this parameter to get the artificial class on the right place.
> 
> However, with my latest commit in [my local repository](https://github.com/vahidrezanezhad/eynollah/tree/postprocess_separators_to_fit_beeter_for_mb_reading_order), you can now obtain textlines without them merging or sticking together.The results are as follows:
> 
> [443594203-c1bae467-7d75-4e28-bb6d-33fdcf9b2c53.zip](https://github.com/user-attachments/files/20205929/443594203-c1bae467-7d75-4e28-bb6d-33fdcf9b2c53.zip)

[443594203-c1bae467-7d75-4e28-bb6d-33fdcf9b2c53.zip](https://github.com/user-attachments/files/20205951/443594203-c1bae467-7d75-4e28-bb6d-33fdcf9b2c53.zip)

**Comment by johnlockejrr** [comment]
> [@johnlockejrr](https://github.com/johnlockejrr) The artificial class can be very helpful based on my experience working with it. However, it is most effective when the ground truth (GT) textlines are not overlapping. This is because the artificial class is generated by dilating the textline contours, and the result is then written onto the label.
> 
> In your case, I notice that the contours overlap significantly. This makes using the artificial class much more complex, as the artificial class for one contour overlaps with its neighboring textline, and vice versa. A better approach would be for the artificial class to be positioned exactly between two textlines, right at their border.
> 
> Given this, I recommend applying [erosion_early](https://github.com/qurator-spk/sbb_pixelwise_segmentation/blob/3de24243fc061d6ef6810515e82cccc4492d7646/gt_gen_utils.py#L688). Unfortunately, this parameter is currently static in the script, but I believe it should be passed as an argument for greater flexibility. You can play with this parameter to get the artificial class on the right place.
> 
> However, with my latest commit in [my local repository](https://github.com/vahidrezanezhad/eynollah/tree/postprocess_separators_to_fit_beeter_for_mb_reading_order), you can now obtain textlines without them merging or sticking together.The results are as follows:
> 
> [443594203-c1bae467-7d75-4e28-bb6d-33fdcf9b2c53.zip](https://github.com/user-attachments/files/20205929/443594203-c1bae467-7d75-4e28-bb6d-33fdcf9b2c53.zip)

The latest commit in your repo for inference with a trained model? How should I use in my case? (what different arguments to pass)

**Comment by johnlockejrr** [comment]
```
            if "artificial_class_label" in keys:
                img_boundary = np.zeros((y_len, x_len))
                erosion_rate = 0#1
                dilation_rate = 2
                dilation_early = 0
                erosion_early = 2
```

Should I increase the `erosion early`?

**Comment by vahidrezanezhad** [comment]
Yes, you need to increase it. Early erosion works by first eroding the contours. By using a higher early_erosion value, you can ensure that the contours no longer overlap. Consequently, the artificial class will be positioned correctly as intended.

**Comment by vahidrezanezhad** [comment]
> By the way, I trained another model (on another dataset) and I'm very satisfied so far (some lines are stuck but very few).
> 
> ![Image](https://github.com/user-attachments/assets/dac7208e-019f-46cb-ae20-9489c8bcc838)
> 
> Could I improve further? I trained the model on `tensorflow==2.19.0` and `keras` v3 but I saved the model in legacy `SaveModel` format with `model.export()` otherwise `keras` doesn't allow to `model.save()` without .keras or .h5 extension, because I wanted to be compatible with `Eynollah`.
> 
> Dataset config:
> 
> ```
> {
>     "use_case": "textline",
>     "artificial_class_label": 2
> }
> ```
> 
> Model config:
> 
> ```
> {
>   "n_classes": 3,
>   "n_epochs": 50,
>   "input_height": 672,
>   "input_width": 448,
>   "weight_decay": 1e-06,
>   "n_batch": 4,
>   "learning_rate": 0.0001,
>   "early_stopping_patience": 15,
>   "early_stopping_monitor": "val_loss",
>   "reduce_lr_patience": 7,
>   "reduce_lr_factor": 0.1,
>   "reduce_lr_min": 1e-07,
>   "reduce_lr_monitor": "val_loss",
>   "patches": false,
>   "augmentation": true,
>   "flip_aug": false,
>   "blur_aug": false,
>   "padding_white": false,
>   "padding_black": false,
>   "scaling": true,
>   "shifting": false,
>   "degrading": false,
>   "brightening": false,
>   "binarization": false,
>   "adding_rgb_background": false,
>   "adding_rgb_foreground": false,
>   "add_red_textlines": false,
>   "channels_shuffling": false,
>   "dir_train": "/home/incognito/sbb_new/dataset/sam_41_mss/dir_train/train",
>   "dir_eval": "/home/incognito/sbb_new/dataset/sam_41_mss/dir_train/eval",
>   "dir_output": "runs/sam_41_mss",
>   "pretraining": true,
>   "scaling_bluring": false,
>   "scaling_binarization": false,
>   "rotation": false,
>   "rotation_not_90": false,
>   "scaling_brightness": false,
>   "scaling_flip": false,
>   "thetha": [
>     10,
>     -10
>   ],
>   "shuffle_indexes": null,
>   "blur_k": [
>     "blur",
>     "guass",
>     "median"
>   ],
>   "scales": [
>     0.6,
>     0.7,
>     0.8,
>     0.9
>   ],
>   "degrade_scales": [
>     0.2,
>     0.4
>   ],
>   "brightness": [
>     1.3,
>     1.5,
>     1.7,
>     2
>   ],
>   "flip_index": [
>     0,
>     1,
>     -1
>   ],
>   "continue_training": false,
>   "transformer_patchsize_x": null,
>   "transformer_patchsize_y": null,
>   "transformer_num_patches_xy": null,
>   "transformer_projection_dim": 64,
>   "transformer_mlp_head_units": [
>     128,
>     64
>   ],
>   "transformer_layers": 8,
>   "transformer_num_heads": 4,
>   "transformer_cnn_first": true,
>   "index_start": 0,
>   "dir_of_start_model": " ",
>   "is_loss_soft_dice": true,
>   "weighted_loss": false,
>   "data_is_provided": false,
>   "task": "segmentation",
>   "f1_threshold_classification": null,
>   "classification_classes_name": null,
>   "backbone_type": "nontransformer",
>   "save_interval": null,
>   "dir_img_bin": null,
>   "number_of_backgrounds_per_image": 1,
>   "dir_rgb_backgrounds": null,
>   "dir_rgb_foregrounds": null,
>   "seed": 691512072
> }
> ```

I think

**Comment by vahidrezanezhad** [comment]
> What can I do in this case? I added the artificial class but I'm not sure is enough for the line not to stick together. Dataset json config:
> 
> ```
> {
>     "use_case": "textline",
>     "artificial_class_label": 2
> }
> ```
> 
> Original image:
> 
> ![Image](https://github.com/user-attachments/assets/c1bae467-7d75-4e28-bb6d-33fdcf9b2c53)
> 
> 3d visualization for the label:
> 
> ![Image](https://github.com/user-attachments/assets/556c2006-0a87-4bb3-9276-1c19402b8a32)
> 
> Aletheia visualization of the textlines:
> 
> ![Image](https://github.com/user-attachments/assets/32c7d67b-e252-4c3a-99b0-a7fdb162ecd8)

If you aim to achieve better and more optimal textline detection for such complex documents, and processing time is not a concern, you can use this [tool](https://github.com/vahidrezanezhad/sbb_standalone_textline_detector). The results are significantly better compared to Eynollah; however, it is extremely slow.

Additionally, you will need high-resolution images since the tool does not automatically adjust the scale.

Results of this tool looks like this

![Image](https://github.com/user-attachments/assets/fef154fa-6ea8-4a26-81c0-212471f11e85)

![Image](https://github.com/user-attachments/assets/263e8069-42be-48c4-a4a9-525fcaac10c4)

[443594203-c1bae467-7d75-4e28-bb6d-33fdcf9b2c53.zip](https://github.com/user-attachments/files/20208106/443594203-c1bae467-7d75-4e28-bb6d-33fdcf9b2c53.zip)

**Comment by johnlockejrr** [comment]
What about the `erosion_rate = 0#1` should I keep it `0` or `1`? I set the `erosion_early = 3`

**Comment by vahidrezanezhad** [comment]
@johnlockejrr Actually, no—set the erosion_rate to zero Focus instead on adjusting early_erosion to determine the best label for your GT. If you want to train the same model as model_hand_ens2022_9^211, you can do so here since it is also a pixel-wise segmentator.

However, the purpose of that model differs from the ones used in Eynollah. The mentioned model's GT represents a thicker baseline with an additional class that marks the exact boundaries between text lines. In contrast, the models in Eynollah are designed for text-line segmentation, not baseline detection—that's the key difference.

By the way, I think you can train a text-line segmentation model with an artificial class that performs exceptionally well on complex images. However, this requires highly accurate GT. The GT must precisely indicate the boundaries between the text lines for the model to succeed.

**Comment by johnlockejrr** [comment]
I trained a model with `erosion_early = 3`.

![Image](https://github.com/user-attachments/assets/33ccbd2c-941c-4235-9969-63178be2031e)

**Comment by vahidrezanezhad** [comment]
@johnlockejrr looks good!

**Comment by johnlockejrr** [comment]
I've trained it with `keras` v3.

https://github.com/johnlockejrr/sbb_pixelwise_segmentation_v3

**Comment by johnlockejrr** [comment]
By the way, the new standalone tool is astonishing better! I still have some problems with the split of some lines but the contours... on not seen data is waaay way better!

![Image](https://github.com/user-attachments/assets/0ee0e525-4ee3-47d7-aa30-d494b8a3de74)

**Comment by vahidrezanezhad** [comment]
> By the way, the new standalone tool is astonishing better! I still have some problems with the split of some lines but the contours... on not seen data is waaay way better!
> 
> ![Image](https://github.com/user-attachments/assets/0ee0e525-4ee3-47d7-aa30-d494b8a3de74)
> 
> Is there a way to finetune dirrectly the `model_hand_ens2022_9^211` on my data (keep the weights) instead of the public resnet50 weights with `sbb_pixelwise_segmentation`? I would also want to finetune it without patches.

Fine-tuning a model makes sense when your training input and output are similar. You can train a new model specifically for standalone text-line detection or fine-tune our existing model. In either case, you will need to prepare your own labels. Try plotting the model_hand_ens2022_9^211 output to visualize how the labels are represented.
